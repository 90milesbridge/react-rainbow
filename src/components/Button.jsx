import React from 'react';
import classnames from 'classnames';

export default class Button extends React.Component {
    render() {
        return (
            <button className={ this.getBtnClass() } 
                    disabled={ this.props.disabled }
                    tabIndex={ this.props.tabIndex }
                    onFocus={ e => this.props.onFocus && this.props.onFocus(e) }
                    onBlur={ e => this.props.onBlur && this.props.onBlur(e) }
                    onClick={ e => this.props.onClick && this.props.onClick(e) }>
                { this.props.children }
            </button>
        )
    }

    getBtnClass() {
        let variant = this.props.variant;
        return classnames('slds-button', {
            'slds-button--neutral': variant === 'neutral',
            'slds-button--brand': variant === 'brand',
            'slds-button--inverse': variant === 'inverse',
            'slds-button--destructive': variant === 'destructive'
        }, this.props.className);
    }
}

Button.PropTypes = {
    variant: React.PropTypes.string,
    disabled: React.PropTypes.bool,
    tabIndex: React.PropTypes.number,
    onClick: React.PropTypes.func,
    onFocus: React.PropTypes.func,
    onBlur: React.PropTypes.func
};

Button.defaultProps = {
    variant: 'neutral',
    disabled: false
};