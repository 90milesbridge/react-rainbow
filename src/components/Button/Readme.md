simple buttons

    <div className="slds-p-vertical_large slds-align--absolute-center slds-wrap" >
        <div className="slds-m-horizontal_medium">
            <Button label="Button Base" />
        </div>
        <div className="slds-m-horizontal_medium">
            <Button label="Button Neutral" variant="neutral" />
        </div>
        <div className="slds-m-horizontal_medium">
            <Button label="Button Outline Brand" variant="outline-brand" />
        </div>
    </div>


button variants

    <div className="slds-p-vertical_large slds-align--absolute-center slds-wrap" >
        <div className="slds-m-horizontal_medium">
            <Button label="Button Brand" onClick={() => alert('clicked!')} variant="brand" />
        </div>
        <div className="slds-m-horizontal_medium">
            <Button label="Button Success" onBlur={ () => alert('blurred!') } variant="success" />
        </div>
        <div className="slds-m-horizontal_medium">
            <Button label="Button Destructive" variant="destructive" />
        </div>
    </div>


buttons with icon

    <div className="slds-p-vertical_large slds-align--absolute-center slds-wrap" >
        <div className="slds-m-horizontal_medium">
            <Button 
                label="Button base"
                iconName="utility:check" 
                iconPosition="left" />
        </div>
        <div className="slds-m-horizontal_medium">
            <Button 
                label="Button with right icon" 
                variant="neutral"
                iconName="utility:forward" 
                iconPosition="right" />
        </div>
        <div className="slds-m-horizontal_medium">
            <Button 
                label="Button with left icon"
                variant="brand"
                iconName="utility:add" 
                iconPosition="left" />
        </div>
    </div>


disabled buttons

    <div className="slds-p-vertical_large slds-align--absolute-center slds-wrap" >
        <div className="slds-m-horizontal_medium">
            <Button label="Button Base Disabled" disabled />
        </div>
        <div className="slds-m-horizontal_medium">
            <Button label="Button Neutral Disabled" variant="neutral" disabled />
        </div>
        <div className="slds-m-horizontal_medium">
            <Button label="Button Brand Disabled" variant="brand" disabled />
        </div>
    </div>


buttons inverse

    const buttonsContainerStyles = {
        backgroundColor: '#16325c',
        borderRadius: '0.25rem',
    };

    <div className="slds-p-vertical_large slds-align--absolute-center" style={buttonsContainerStyles}>
        <div className="slds-m-horizontal_medium">
            <Button label="Button Inverse" variant="inverse" />
        </div>
        <div className="slds-m-horizontal_medium">
            <Button 
                label="Button Inverse with icon"
                variant="inverse"
                iconName="utility:add" 
                iconPosition="left" />
        </div>
    </div>
