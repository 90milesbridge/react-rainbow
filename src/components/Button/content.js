
export default function Content({ label, children }) {
    if (children.length > 0) {
        return children;
    }
    return label;
}
