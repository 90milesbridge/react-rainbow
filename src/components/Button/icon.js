import React from 'react';
import PropTypes from 'prop-types';
import IconSvg from '../IconSvg';

function ButtonIcon(props) {
    const { isVisible, iconName, position } = props;

    const getIconClassName = () => `slds-button__icon slds-button__icon_${position}`;

    if (isVisible) {
        return <IconSvg iconName={iconName} className={getIconClassName()} />;
    }
    return null;
}

ButtonIcon.propTypes = {
    isVisible: PropTypes.bool.isRequired,
    iconName: PropTypes.string.isRequired,
    position: PropTypes.string.isRequired,
};

export default ButtonIcon;
