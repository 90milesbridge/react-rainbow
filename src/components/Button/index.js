import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import iconNameUtilityPropType from './../../propTypes/iconNameUtilityPropType';
import LeftIcon from './leftIcon';
import RightIcon from './rightIcon';

/**
* Buttons are clickable items used to perform an action.
*/
export default class Button extends Component {
    constructor(props) {
        super(props);
        this.buttonRef = React.createRef();
        this.click = this.click.bind(this);
        this.focus = this.focus.bind(this);
        this.blur = this.blur.bind(this);
    }

    getClassNames() {
        const { className, variant } = this.props;
        return classnames('slds-align_absolute-center', 'slds-button', `slds-button_${variant}`, className);
    }

    /**
    * Sets focus on the element.
    * @public
    */
    focus() {
        this.buttonRef.current.focus();
    }

    /**
    * Sets click on the element.
    * @public
    */
    click() {
        this.buttonRef.current.click();
    }

    /**
    * Sets blur on the element.
    * @public
    */
    blur() {
        this.buttonRef.current.blur();
    }

    render() {
        const {
            style,
            label,
            disabled,
            tabIndex,
            onFocus,
            onBlur,
            onClick,
            title,
            type,
            ariaHaspopup,
            iconName,
            iconPosition,
        } = this.props;

        return (
            <button
                data-id="button-element"
                className={this.getClassNames()}
                style={style}
                disabled={disabled}
                tabIndex={tabIndex}
                onFocus={onFocus}
                onBlur={onBlur}
                onClick={onClick}
                title={title}
                type={type}
                aria-haspopup={ariaHaspopup}
                ref={this.buttonRef} >

                <LeftIcon iconName={iconName} position={iconPosition} />
                {label}
                <RightIcon iconName={iconName} position={iconPosition} />
            </button>
        );
    }
}

Button.propTypes = {
    /** The text to be displayed inside the button. */
    label: PropTypes.oneOfType([
        PropTypes.string, PropTypes.node]).isRequired,
    /** The variant changes the appearance of the button.
    * Accepted variants include base, neutral, brand, outline-brand, destructive,
    * success and inverse. This value defaults to base. */
    variant: PropTypes.oneOf([
        'base',
        'neutral',
        'brand',
        'outline-brand',
        'destructive',
        'success',
        'inverse',
    ]),
    /** The Lightning Design System name of the icon used as a fallback when
    * the image fails to load. Names are written in the format {sprite_name}:{icon_name}
    * where {sprite_name} is the category, and {icon_name} is the specific icon to be displayed.
    * Only utility icons can be used in this component. */
    iconName: iconNameUtilityPropType,
    /** Describes the position of the icon with respect to body. Options include left and right.
    * This value defaults to left. */
    iconPosition: PropTypes.oneOf([
        'left', 'right',
    ]),
    /** Displays tooltip text when the mouse moves over the element. */
    title: PropTypes.string,
    /** Specifies the type of button. Valid values are button, reset, and submit.
    * This value defaults to button. */
    type: PropTypes.oneOf([
        'button', 'submit', 'reset',
    ]),
    /** Specifies whether this button should be displayed in a disabled state.
    * Disabled buttons can't be clicked. This value defaults to false. */
    disabled: PropTypes.bool,
    /** Specifies the tab order of an element (when the tab button is used for navigating). */
    tabIndex: PropTypes.number,
    /** The action triggered when the button is clicked. */
    onClick: PropTypes.func,
    /** The action triggered when the element receives focus. */
    onFocus: PropTypes.func,
    /** The action triggered when the element releases focus. */
    onBlur: PropTypes.func,
    /** Indicates that the element has a popup context menu or sub-level menu. */
    ariaHaspopup: PropTypes.bool,
    /** A CSS class for the outer element, in addition to the component's base classes. */
    className: PropTypes.string,
    /** An object with custom style applied to the outer element. */
    style: PropTypes.object,
};

Button.defaultProps = {
    variant: 'base',
    iconName: '',
    iconPosition: 'left',
    title: undefined,
    type: 'button',
    disabled: false,
    tabIndex: undefined,
    onClick: () => {},
    onFocus: () => {},
    onBlur: () => {},
    ariaHaspopup: false,
    className: undefined,
    style: undefined,
};
