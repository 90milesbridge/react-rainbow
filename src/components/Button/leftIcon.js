import React from 'react';
import PropTypes from 'prop-types';
import IconSvg from '../IconSvg';

export default function LeftIcon({ iconName, position }) {
    if (iconName && position === 'left') {
        return (
            <IconSvg iconName={iconName} className="slds-button__icon slds-button__icon_left" />
        );
    }
    return null;
}

LeftIcon.propTypes = {
    iconName: PropTypes.string.isRequired,
    position: PropTypes.string.isRequired,
};
