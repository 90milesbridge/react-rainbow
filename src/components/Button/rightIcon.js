import React from 'react';
import PropTypes from 'prop-types';

export default function RightIcon({ icon, position }) {
    if (icon && position === 'right') {
        return (
            <span className="slds-button__icon slds-button__icon_right">
                {icon}
            </span>
        );
    }
    return null;
}

RightIcon.propTypes = {
    icon: PropTypes.node.isRequired,
    position: PropTypes.oneOf([
        'left', 'right',
    ]),
};

RightIcon.defaultProps = {
    position: undefined,
};
