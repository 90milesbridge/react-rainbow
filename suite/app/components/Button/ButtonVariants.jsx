import React from 'react';
import { Button } from './../../../../entry';

export default class ButtonVariants extends React.Component {
    render() {
        return (
            <div>
                <Button variant="base">Base</Button>
                <Button variant="neutral">Neutral</Button>
                <Button variant="brand">Brand</Button>
                <Button variant="destructive">Destructive</Button>
            </div>
        )
    }
} 