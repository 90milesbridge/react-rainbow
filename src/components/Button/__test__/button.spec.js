import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Button from './../index';

describe('<Button/>', () => {
    it('should call onClick function when someone click over', () => {
        const onClickMockFn = jest.fn();
        const component = mount(
            <Button label="Button label" onClick={onClickMockFn} />,
        );

        component.simulate('click');
        expect(onClickMockFn.mock.calls.length).toBe(1);
    });
    it('should call onBlur function when it lost the focus', () => {
        const onBlurMockFn = jest.fn();
        const component = mount(
            <Button label="Button label" onBlur={onBlurMockFn} />,
        );

        component.simulate('blur');
        expect(onBlurMockFn.mock.calls.length).toBe(1);
    });
    it('should call onFocus function when it gets the focus', () => {
        const onFocusMockFn = jest.fn();
        const component = mount(
            <Button label="Button label" onFocus={onFocusMockFn} />,
        );

        component.simulate('focus');
        expect(onFocusMockFn.mock.calls.length).toBe(1);
    });
    it('should be defined the click method', () => {
        const component = mount(
            <Button />,
        );

        expect(typeof component.instance().click).toBe('function');
    });
    it('should be defined the focus method', () => {
        const component = mount(
            <Button />,
        );

        expect(typeof component.instance().focus).toBe('function');
    });
    it('should be defined the blur method', () => {
        const component = mount(
            <Button />,
        );

        expect(typeof component.instance().blur).toBe('function');
    });
    it('should focus the button when the focus method is called', () => {
        const component = mount(
            <Button />,
        );

        component.instance().focus();
        const focusedElementDataId = document.activeElement.getAttribute('data-id');
        const buttonDataId = component.find('button').prop('data-id');
        expect(focusedElementDataId).toBe(buttonDataId);
    });
    it('should blur the button when the blur method is called', () => {
        const component = mount(
            <Button />,
        );
        const instance = component.instance();
        const buttonDataId = component.find('button').prop('data-id');

        instance.focus();
        expect(document.activeElement.getAttribute('data-id')).toBe(buttonDataId);
        instance.blur();
        expect(document.activeElement.getAttribute('data-id')).toBeNull();
    });
    it('should set disable to true when it is passed', () => {
        const component = mount(
            <Button label="my label" disabled />,
        );

        expect(component.find('button').prop('disabled')).toBe(true);
    });
    it('should set the tabIndex passed', () => {
        const component = mount(
            <Button label="my label" tabIndex={0} />,
        );

        expect(component.find('button').prop('tabIndex')).toBe(0);
    });
    it('should set the title passed', () => {
        const component = mount(
            <Button label="my label" title="my title" />,
        );

        expect(component.find('button').prop('title')).toBe('my title');
    });
    it('should set the type passed', () => {
        const component = mount(
            <Button label="my label" type="reset" />,
        );

        expect(component.find('button').prop('type')).toBe('reset');
    });
    it('should set aria-haspopup to true when it is passed', () => {
        const component = mount(
            <Button label="my label" ariaHaspopup />,
        );

        expect(component.find('button').prop('aria-haspopup')).toBe(true);
    });
    it('should set the label passed as children', () => {
        const component = mount(
            <Button label="Click me" />,
        );

        expect(component.find('button').text()).toBe('Click me');
    });
    it('should have the right class names', () => {
        const component = renderer.create(
            <Button label="Button Label" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant neutral and have a custom class', () => {
        const component = renderer.create(
            <Button label="Button Label" variant="neutral" className="my-custom-class-name" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant brand', () => {
        const component = renderer.create(
            <Button label="Button Label" variant="brand" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant outline-brand', () => {
        const component = renderer.create(
            <Button label="Button Label" variant="outline-brand" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant destructive', () => {
        const component = renderer.create(
            <Button label="Button Label" variant="destructive" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant success', () => {
        const component = renderer.create(
            <Button label="Button Label" variant="success" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant inverse', () => {
        const component = renderer.create(
            <Button label="Button Label" variant="inverse" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should render an icon on the left when iconName is passed', () => {
        const component = mount(
            <Button label="Button Label" iconName="utility:world" />,
        );
        const leftIcon = component.find('[data-id="left-icon"]');
        const rightIcon = component.find('[data-id="right-icon"]');

        expect(leftIcon.prop('isVisible')).toBe(true);
        expect(rightIcon.prop('isVisible')).toBe(false);
    });
    it('should render an icon on the right when iconName is passed and iconPosition is right', () => {
        const component = mount(
            <Button label="Button Label" iconName="utility:world" iconPosition="right" />,
        );
        const leftIcon = component.find('[data-id="left-icon"]');
        const rightIcon = component.find('[data-id="right-icon"]');

        expect(leftIcon.prop('isVisible')).toBe(false);
        expect(rightIcon.prop('isVisible')).toBe(true);
    });
});
