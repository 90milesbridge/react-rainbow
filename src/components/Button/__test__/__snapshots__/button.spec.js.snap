// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should have the right class names 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_base"
  data-id="button-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  Button Label
</button>
`;

exports[`<Button/> should have the right class names when variant brand 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_brand"
  data-id="button-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  Button Label
</button>
`;

exports[`<Button/> should have the right class names when variant destructive 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_destructive"
  data-id="button-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  Button Label
</button>
`;

exports[`<Button/> should have the right class names when variant inverse 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_inverse"
  data-id="button-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  Button Label
</button>
`;

exports[`<Button/> should have the right class names when variant neutral and have a custom class 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_neutral my-custom-class-name"
  data-id="button-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  Button Label
</button>
`;

exports[`<Button/> should have the right class names when variant outline-brand 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_outline-brand"
  data-id="button-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  Button Label
</button>
`;

exports[`<Button/> should have the right class names when variant success 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_success"
  data-id="button-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  Button Label
</button>
`;
